### PR TITLE
[ISSUE] add filteredMsgCount for `pulsar-admin topics stats`

### PR DESCRIPTION
**Pull request by tjiuming**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*
Fixes: #14507 

### Motivation
see master issue
### Modifications
1. add field `msgFilteredCounter` in `AbstractTopic`
2. add methods `addFilteredMsgCount`, `getMsgFilteredCount` in `AbstractTopic`
3. record filtered message count in `AbstractBaseDispatcher#filterEntriesForConsumer`

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below or label this PR directly (if you have committer privilege).

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Comment by tjiuming** [comment]
tests to to completed.

**Review by codelipenghui — COMMENTED** [review]
LGTM, please add tests.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java [hunk 1]
```diff
@@ -220,6 +220,12 @@ public int filterEntriesForConsumer(Optional<EntryWrapper[]> entryWrapper, int e
         if (CollectionUtils.isNotEmpty(entriesToFiltered)) {
             subscription.acknowledgeMessage(entriesToFiltered, AckType.Individual,
                     Collections.emptyMap());
+
+            int filtered = entriesToFiltered.size();
```
Comment (Jason918) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java:224]: It's actually entry count, not message count. Please make it consist.

Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java:224]: make sense, I'll change the field name. Thanks!

**Review by Jason918 — APPROVED** [review]

**Review by gaoran10 — APPROVED** [review]
LGTM

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/plugin/FilterEntryTest.java [hunk 2]
```diff
@@ -159,4 +155,52 @@ public void testFilter() throws Exception {
 
     }
 
+
+    @Test
+    public void testFilteredMsgCount() throws Throwable{
+        String topic = "persistent://prop/ns-abc/topic" + UUID.randomUUID();
+        String subName = "sub";
+        Consumer<String> consumer = pulsarClient.newConsumer(Schema.STRING).topic(topic)
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/plugin/FilterEntryTest.java:163]:
```suggestion
        @Cleanup
        Consumer<String> consumer = pulsarClient.newConsumer(Schema.STRING).topic(topic)
```

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/plugin/FilterEntryTest.java [hunk 3]
```diff
@@ -159,4 +155,52 @@ public void testFilter() throws Exception {
...
+        Field field = AbstractBaseDispatcher.class.getDeclaredField("entryFilters");
+        field.setAccessible(true);
+        NarClassLoader narClassLoader = mock(NarClassLoader.class);
+        EntryFilter filter1 = new EntryFilterTest();
+        EntryFilterWithClassLoader loader1 = spyWithClassAndConstructorArgs(EntryFilterWithClassLoader.class, filter1, narClassLoader);
+        EntryFilter filter2 = new EntryFilter2Test();
+        EntryFilterWithClassLoader loader2 = spyWithClassAndConstructorArgs(EntryFilterWithClassLoader.class, filter2, narClassLoader);
+        field.set(dispatcher, ImmutableList.of(loader1, loader2));
+
+        Producer<String> producer = pulsarClient.newProducer(Schema.STRING)
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/plugin/FilterEntryTest.java:178]:
```suggestion
        @Cleanup
        Producer<String> producer = pulsarClient.newProducer(Schema.STRING)
```

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/plugin/FilterEntryTest.java [hunk 4]
```diff
@@ -159,4 +155,52 @@ public void testFilter() throws Exception {
...
+        }
+
+        for (int i = 0; i < 10; i++) {
+            assertNotNull(producer.newMessage().property("REJECT", "").value("1").send());
+        }
+
+
+        int counter = 0;
+        while (true) {
+            Message<String> message = consumer.receive(1, TimeUnit.SECONDS);
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/plugin/FilterEntryTest.java:191]: 1 second might be too short in the CI environment if the broker dispatch messages slowly.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/plugin/FilterEntryTest.java [hunk 5]
```diff
@@ -32,10 +31,7 @@
 import java.util.UUID;
 import java.util.concurrent.TimeUnit;
 import org.apache.bookkeeper.mledger.impl.PositionImpl;
-import org.apache.pulsar.broker.service.AbstractBaseDispatcher;
-import org.apache.pulsar.broker.service.BrokerService;
-import org.apache.pulsar.broker.service.BrokerTestBase;
-import org.apache.pulsar.broker.service.Dispatcher;
+import org.apache.pulsar.broker.service.*;
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/plugin/FilterEntryTest.java:34]: Please avoid star import.

**Comment by dao-jun** [comment]
@codelipenghui  changed, PTAL

**Review by codelipenghui — APPROVED** [review]

**Comment by michaeljmarshall** [comment]
@tjiuming - thank you for adding these stats. Would you be able to add the documentation to the admin api docs https://pulsar.apache.org/docs/next/admin-api-topics#get-stats?

**Comment by michaeljmarshall** [comment]
Note that this PR did not add these metrics to the `pulsar-admin topics stats` command. It added it to the `pulsar-admin broker-stats topics` command. It would be helpful to expose these metrics on the `topics stats` command as subscription level metrics. It'd also be nice to add it as the prometheus metrics output.

**Comment by tjiuming** [comment]
> Note that this PR did not add these metrics to the `pulsar-admin topics stats` command. It added it to the `pulsar-admin broker-stats topics` command. It would be helpful to expose these metrics on the `topics stats` command as subscription level metrics. It'd also be nice to add it as the prometheus metrics output.

All right

**Pull request merged**